### PR TITLE
Adding support for FP16 Alt Round Near Zero Mode

### PR DESCRIPTION
**Pull request by yoichiyoshida**
Current FP16 Alt implementation truncates conversions between FP32 and BF16 in the intermediate step. This PR adds support for a second FP16 Alt implementation which rounds from FP32 to BF16.

### CODE REVIEW

**Comment by nakajee** [comment]
What happens if we specify Fp16AltImpl=False and Fp16AltImplRound=True?
Rejected? Or force Fp16AltImplRound=False?

**Review by AlexBrownAMD — COMMENTED** [review]
Looks like this is missing the changes to add the new predicates to the library logic. Not needed often, but if the predicates are missed at some point it could silently cause bad selections and numerical errors.

In LibrarioIO.py createLibraryLogicList:

if "Fp16AltImpl" in problemType and problemType["Fp16AltImpl"]:
  data.append("Fp16AltImpl")
  if "Fp16AltImplRound" in problemType and problemType["Fp16AltImplRound"]:
    data.append("Fp16AltImplRound")

File: Tensile/SolutionStructs.py [hunk 1]
```diff
@@ -1141,7 +1141,9 @@ def __str__(self):
     # Other
     if self["UseBeta"]: name += "B"
     if self["HighPrecisionAccumulate"] and not self["SilentHighPrecisionAccumulate"]: name += "H"
-    if self["Fp16AltImpl"]: name += "R"
+    if self["Fp16AltImpl"]:
```
[the remark this answers is not in the document]
Comment (yoichiyoshida) [reply Tensile/SolutionStructs.py:1144]: If FP16AltImpl is set we need to see R(Z). If there are improper cases it needs to not be hidden with something like this.

File: Tensile/Tests/pre_checkin/denorm/mfma/hgemm_denorm_alt.yaml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      Batched: True
+      Fp16AltImpl: True
+
+    - # BenchmarkProblemSizeGroup - Standard
+      InitialSolutionParameters:
+      BenchmarkCommonParameters:
+        - KernelLanguage: ["Assembly"]
+        - EdgeType: ["ShiftPtr"]
+      ForkParameters:
+        - TransposeLDS: [0, 1]
```
[the remark this answers is not in the document]
Comment (yoichiyoshida) [reply Tensile/Tests/pre_checkin/denorm/mfma/hgemm_denorm_alt.yaml:38]: This is a copy paste of the regular denorm case but for fp16altround = 1. Needs to test that options are not broken with the addition

**Comment by yoichiyoshida** [comment]
> What happens if we specify Fp16AltImpl=False and Fp16AltImplRound=True? Rejected? Or force Fp16AltImplRound=False?

Round only affects the kernel type generated for FP16Alt, if Alt is disabled we are generating regular kernels.

**Comment by nakajee** [comment]
> > What happens if we specify Fp16AltImpl=False and Fp16AltImplRound=True? Rejected? Or force Fp16AltImplRound=False?
> 
> Round only affects the kernel type generated for FP16Alt, if Alt is disabled we are generating regular kernels.

OK. I quickly tried Fp16AltImpl=False and Fp16AltImplRound=True.
It generates same kernel as both False

**Review by AlexBrownAMD — APPROVED** [review]

**Review by nakajee — APPROVED** [review]

**Pull request merged**